README.md

```diff
@@ -1,116 +1,267 @@
+
+To prepare a YAML config file, copy [template_config.yaml](./configs/template_config.yaml) into the [user_configs](./configs/user_configs/) folder. Fill out all parameters accordingly. Absolute paths are preferred for any path variables. The repository is setup to work flexibly with any desired directory structure.
 
-#### 1. **Running a Training Script with Hyperparameters**:
+### Execution Workflow
 
-Assuming you have a training script `train.py`, you can run it with different hyperparameters.
+Once a YAML config file is prepared, you can pass this file into any script in the pipeline. Before you run any scripts, it is recommended to copy all needed scripts into the [user_scripts](./scripts/user_scripts/) folder and modify the scripts to point to the right config file.
+
+The expected order of script execution is as follows:
+1. Run `download_data.sh` to fetch the necessary data.
+2. Execute `train_tokenizer.sh` to prepare the tokenizer.
+3. Use `tokenize_data.sh` for data tokenization.
+4. Finally, run `retnet.sh` or `transformer.sh`, depending on your project needs.
+
+For example, if you want to train a RetNet model:
 
 ```bash
-python train.py --learning-rate 0.001 --batch-size 64 --num-layers 4 --embed-dim 256
+./download_data.sh
+./train_tokenizer.sh
+./tokenize_data.sh
+
+# Replace this with the model that you want to run
+./retnet.sh
 ```
 
-#### 2. **Using a Configuration File**:
+### Note on CUDA
+
+Nvidia GPUs are recommended to speed up computation, though a CPU can alternatively be used. Nvidia drivers and the CUDA toolkit will also need to be installed. To test if there is an Nvidia GPU available on your system, see instructions for [Linux](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#verify-you-have-a-cuda-capable-gpu) and [Windows](https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html#verify-you-have-a-cuda-capable-gpu).
+
+## Features
 
-For complex models, managing hyperparameters via a config file can be more convenient.
+### Grid Search
+
+The Grid Search feature is designed to systematically explore a range of hyperparameters and compare RetNet and Transformer models with corresponding parameters at each point. This evaluates both architectures with various combinations of learning rates, embedding dimensions, feed-forward dimensions, sequence lengths, and number of heads. The goal is to identify the configuration that results in the best model performance, measured in terms of loss and training efficiency.
+
+**Code Overview:**
+
+We implement the grid search process as follows:
+
+- **Hyperparameters Tested:** Learning rates (`0.001`, `0.0005`, `0.0001`), embedding dimensions (`768`, `1024`, `1280`), feed-forward dimensions (`1024`, `2048`), heads (`4`, `8`), and sequence lengths (`256`, `512`) for a total of 72 unique combinations.
```

Comment:
You mean refer to the default yaml file?

Reply:
@RLin8103, this was pre-yaml, and wouldn't be good for a yaml anyway because we'd be referening 144 different yamls. I think he means reference them to the actual grid_search.sh that we ran, but it isn't compatible with our current code base, so I think pointing users to it would be misleading. IMO, this is the easist way to communicate that info unless it's in the paper (appendix). Until it is, I feel like it's fine to leave it here.